bokehjs/src/lib/models/glyphs/vsteparea.ts

```diff
@@ -0,0 +1,184 @@
+    const {min, max} = Math
+
+    for (let i = 0; i < this.data_size; i++) {
+      const x = this._x[i]
+      const y1 = this._y1[i]
+      const y2 = this._y2[i]
+      index.add_rect(x, min(y1, y2), x, max(y1, y2))
+    }
+
+  }
+
+  protected _step_path(ctx: Context2d, mode: StepMode, sx: ScreenArray, sy: ScreenArray, from_i: number, to_i: number): void {
+    // Assume that path was already moved to the first point
+    let prev_x = sx[from_i]
+    let prev_y = sy[from_i]
+    const idx_dir = from_i < to_i ? 1 : -1
+    for (let i = from_i + idx_dir; i != to_i; i += idx_dir) {
+      switch (mode) {
+        case "before":
+          ctx.lineTo(prev_x, sy[i])
+          ctx.lineTo(sx[i], sy[i])
+          break
+        case "after":
+          ctx.lineTo(sx[i], prev_y)
+          ctx.lineTo(sx[i], sy[i])
+          break
+        case "center":
+          const mid_x = (prev_x + sx[i]) / 2
+          ctx.lineTo(mid_x, prev_y)
+          ctx.lineTo(mid_x, sy[i])
+          ctx.lineTo(sx[i], sy[i])
+          break
+        default:
+          unreachable()
+      }
+      prev_x = sx[i]
+      prev_y = sy[i]
+    }
+  }
+
+  protected _flip_mode(mode: StepMode): StepMode {
+    switch (mode) {
+      case "before":
+        return "after"
+      case "after":
+        return "before"
+      case "center":
+        return "center"
+      default:
+        unreachable()
+    }
+  }
+
+  protected _render(ctx: Context2d, _indices: number[], data?: VStepAreaData): void {
+    const {sx, sy1, sy2} = data ?? this
+
+    const forward_mode = this.model.step_mode
+    const backward_mode = this._flip_mode(this.model.step_mode)
+
+    ctx.beginPath()
+    ctx.moveTo(sx[0], sy1[0])
+
+    this._step_path(ctx, forward_mode, sx, sy1, 0, sx.length)
+    this._step_path(ctx, backward_mode, sx, sy2, sx.length, -1)
+
+    ctx.closePath()
+
+    this.visuals.fill.apply(ctx)
+    this.visuals.hatch.apply(ctx)
+  }
+
+  scenterxy(i: number): [number, number] {
+    const scx = this.sx[i]
+    const scy = (this.sy1[i] + this.sy2[i])/2
+    return [scx, scy]
+  }
+
+  protected override _hit_point(geometry: PointGeometry): Selection {
+    const {sx, sy1, sy2} = this
+
+    for (let i = 0; i < this.sx.length - 1; i++) {
+      let px: number[]
+      let py: number[]
+
+      switch (this.model.step_mode) {
+        case "before":
+          px = [sx[i], sx[i+1], sx[i+1], sx[i]]
+          py = [sy1[i+1], sy1[i+1], sy2[i+1], sy2[i+1]]
+          break
+        case "after":
+          px = [sx[i], sx[i+1], sx[i+1], sx[i]]
+          py = [sy1[i], sy1[i], sy2[i], sy2[i]]
+          break
+        case "center":
+          const mid_x = (sx[i] + sx[i+1]) / 2
+          px = [sx[i], mid_x, mid_x, sx[i+1], sx[i+1], mid_x, mid_x, sx[i]]
+          py = [sy1[i], sy1[i], sy1[i+1], sy1[i+1], sy2[i+1], sy2[i+1], sy2[i], sy2[i]]
+          break
+        default:
+          console.warn("step_mode", this.model.step_mode)
+          unreachable()
```

Comment:
This switch is already exhaustive, which the compiler can verify. `unreachable()` is meant for situations where tsc can't prove exhaustivity for some reason. This default case can be safely removed.

Reply:
Removed it here as well as in the other switches. I'm curious though about why I see no tsc/linter complains if done wrong, but we can discuss this in Slack if you like.

Reply:
This actually a feature of TypeScript. Its goal is to augment JavaScript, not to replace it. So it doesn't complain about the default case, because in some scenarios devs will use types, but given that consumers will likely use the APIs from raw JS, then there is a possibility for giving those users a useful error message. We, however, rely heavily on TS and there's already a layer of validation in the model, so having additional checking here is unnecessary, and would needlessly bloat generated JS bundles. However, having a lint rule for this would be useful.

Reply:
There are other features like this, e.g. nullability checks for non-nullable types (even under `strictNullChecks: true`).

Reply:
Sorry, I didn't elaborate what I meant: I'm okay with there not being a "pointless default case in exhaustive switch" complaint. I was missing a complaint if the switch is _not_ exhaustive and has _no_ default case either. The idea being that every switch like this would alert you about a missing case if you ever extend `StepMode`.

Reply:
Right, this is the one case (or maybe one of cases) where TypeScript actually doesn't help with exhaustivity checking. So in PR #12853 I fixed that issue, so now at least we will get a validation error for non-exhaustive switches.